staker/challenge-cache/cache.go

```diff
@@ -0,0 +1,242 @@
+	ErrNoHashes           = errors.New("no hashes being written")
+	hashesFileName        = "hashes.bin"
+	wavmModuleRootPrefix  = "wavm-module-root"
+	rollupBlockHashPrefix = "rollup-block-hash"
+	messageNumberPrefix   = "message-num"
+	bigStepPrefix         = "big-step"
+	challengeLevelPrefix  = "subchallenge-level"
+)
+
+// HistoryCommitmentCacher can retrieve history commitment hashes given lookup keys.
+type HistoryCommitmentCacher interface {
+	Get(lookup *Key, numToRead uint64) ([]common.Hash, error)
+	Put(lookup *Key, hashes []common.Hash) error
+}
+
+// Cache for history commitments on disk.
+type Cache struct {
+	baseDir string
+}
+
+// New cache from a base directory path.
+func New(baseDir string) (*Cache, error) {
+	if _, err := os.Stat(baseDir); err != nil {
+		if err := os.MkdirAll(baseDir, os.ModePerm); err != nil {
+			return nil, fmt.Errorf("could not make base cache directory %s: %w", baseDir, err)
+		}
+	}
+	return &Cache{
+		baseDir: baseDir,
+	}, nil
+}
+
+// Key for cache lookups includes the wavm module root of a challenge, as well
+// as the heights for messages and big steps as needed.
+type Key struct {
+	RollupBlockHash common.Hash
+	WavmModuleRoot  common.Hash
+	MessageHeight   uint64
+	StepHeights     []uint64
+}
+
+// Get a list of hashes from the cache from index 0 up to a certain index. Hashes are saved as files in the directory
+// hierarchy for the cache. If a file is not present, ErrNotFoundInCache
+// is returned.
+func (c *Cache) Get(
+	lookup *Key,
+	numToRead uint64,
+) ([]common.Hash, error) {
+	fName, err := determineFilePath(c.baseDir, lookup)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := os.Stat(fName); err != nil {
+		log.Warn("Cache miss", "fileName", fName)
+		return nil, ErrNotFoundInCache
+	}
+	log.Debug("Cache hit", "fileName", fName)
+	f, err := os.Open(fName)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if err := f.Close(); err != nil {
+			log.Error("Could not close file after reading", "err", err, "file", fName)
+		}
+	}()
+	return readHashes(f, numToRead)
+}
+
+// Put a list of hashes into the cache.
+// Hashes are saved as files in a directory hierarchy for the cache.
+// This function first creates a temporary file, writes the hashes to it, and then renames the file
+// to the final directory to ensure atomic writes.
+func (c *Cache) Put(lookup *Key, hashes []common.Hash) error {
+	// We should error if trying to put 0 hashes to disk.
+	if len(hashes) == 0 {
+		return ErrNoHashes
+	}
+	fName, err := determineFilePath(c.baseDir, lookup)
+	if err != nil {
+		return err
+	}
+	// We create a tmp file to write our hashes to first. If writing fails,
+	// we don't want to leave a half-written file in our cache directory.
+	// Once writing succeeds, we rename in an atomic operation to the correct file name
+	// in the cache directory hierarchy.
+	tmp, err := os.MkdirTemp(c.baseDir, "tmpdir")
+	if err != nil {
+		return err
+	}
+	tmpFName := filepath.Join(tmp, fName)
```

Comment:
I think that should be determineFilePath(tmp, lookup).


As it is, seems like you have the base-path twice in there?